src/librbd/mirror/snapshot/UnlinkPeerRequest.cc

```diff
@@ -91,15 +91,20 @@ void UnlinkPeerRequest<I>::unlink_peer() {
     finish(-EINVAL);
     return;
   }
-
-  // if there is or will be no more peers in the mirror snapshot and we have
-  // a more recent mirror snapshot, remove the older one
   if ((mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) == 0) ||
       (mirror_ns->mirror_peer_uuids.size() <= 1U && m_newer_mirror_snapshots)) {
-    m_image_ctx->image_lock.unlock_shared();
-    remove_snapshot();
-    return;
+
+    if (m_allow_remove) {
+      m_image_ctx->image_lock.unlock_shared();
+      remove_snapshot();
+      return;
+    } else if ((mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) == 0)) {
```

Comment:
nit: as the upper branch has `return` statement, I would use just `if` (not `else if`) to be consistent with the previous code. Though it is just my preference and I will be fine with the current version too.